File: Comprehensive Health Data Analytics for Chronic Disease Prediction/python_scripts/comprehensive_eda.py
"""
Comprehensive Exploratory Data Analysis
Deep dive into the health dataset
"""

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
import missingno as msno
import warnings
warnings.filterwarnings('ignore')

sns.set_style('whitegrid')
plt.rcParams['figure.figsize'] = (14, 8)

def load_and_clean_data():
    """Load and clean the health data"""
    df = pd.read_csv('../data/health_data.csv')
    if 'Unnamed: 0' in df.columns:
        df = df.drop('Unnamed: 0', axis=1)
    df['age_years'] = df['age'] / 365.25
    df['bmi'] = df['weight'] / ((df['height'] / 100) ** 2)
    df = df[(df['ap_hi'] >= 80) & (df['ap_hi'] <= 250)]
    df = df[(df['ap_lo'] >= 40) & (df['ap_lo'] <= 150)]
    df = df[df['ap_hi'] >= df['ap_lo']]
    df = df[(df['height'] >= 100) & (df['height'] <= 220)]
    df = df[(df['weight'] >= 30) & (df['weight'] <= 200)]
    df = df[(df['bmi'] >= 10) & (df['bmi'] <= 60)]
    return df

def data_overview(df):
    """Provide comprehensive data overview"""
    print("=" * 80)
    print("DATA OVERVIEW")
    print("=" * 80)
    print(f"Dataset shape: {df.shape}")
    print(f"\nColumn names: {df.columns.tolist()}")
    print(f"\nData types:\n{df.dtypes}")
    print(f"\nMissing values:\n{df.isnull().sum()}")
    print(f"\nDuplicate rows: {df.duplicated().sum()}")
    print(f"\nFirst few rows:")
    print(df.head())
    print(f"\nSummary statistics:")
    print(df.describe())
    
    # Missing data visualization
    if df.isnull().sum().sum() > 0:
        msno.matrix(df)
        plt.title('Missing Data Pattern', fontsize=14, fontweight='bold')
        plt.tight_layout()
        plt.savefig('../figures/missing_data.png', dpi=300, bbox_inches='tight')
        plt.close()

def outlier_analysis(df):
    """Analyze outliers in the dataset"""
    print("=" * 80)
    print("OUTLIER ANALYSIS")
    print("=" * 80)
    
    numerical_cols = ['age_years', 'height', 'weight', 'ap_hi', 'ap_lo', 'bmi']
    fig, axes = plt.subplots(2, 3, figsize=(18, 12))
    axes = axes.ravel()
    
    for i, col in enumerate(numerical_cols):
        Q1 = df[col].quantile(0.25)
        Q3 = df[col].quantile(0.75)
        IQR = Q3 - Q1
        lower_bound = Q1 - 1.5 * IQR
        upper_bound = Q3 + 1.5 * IQR
        
        outliers = df[(df[col] < lower_bound) | (df[col] > upper_bound)]
        print(f"\n{col.upper()}:")
        print(f"  Outliers: {len(outliers)} ({len(outliers)/len(df)*100:.2f}%)")
        print(f"  Lower bound: {lower_bound:.2f}")
        print(f"  Upper bound: {upper_bound:.2f}")
        
        axes[i].boxplot(df[col], vert=True)
        axes[i].set_title(f'{col.upper()} - Box Plot', fontsize=12, fontweight='bold')
        axes[i].set_ylabel(col)
        axes[i].grid(True, alpha=0.3, axis='y')
    
    plt.tight_layout()
    plt.savefig('../figures/outlier_analysis.png', dpi=300, bbox_inches='tight')
    plt.close()

def distribution_analysis(df):
    """Analyze distributions of variables"""
    print("=" * 80)
    print("DISTRIBUTION ANALYSIS")
    print("=" * 80)
    
    numerical_cols = ['age_years', 'height', 'weight', 'ap_hi', 'ap_lo', 'bmi']
    fig, axes = plt.subplots(2, 3, figsize=(18, 12))
    axes = axes.ravel()
    
    for i, col in enumerate(numerical_cols):
        # Histogram with normal distribution overlay
        axes[i].hist(df[col], bins=50, density=True, alpha=0.7, edgecolor='black', label='Data')
        
        # Normal distribution
        mu, sigma = df[col].mean(), df[col].std()
        x = np.linspace(df[col].min(), df[col].max(), 100)
        axes[i].plot(x, stats.norm.pdf(x, mu, sigma), 'r-', linewidth=2, label='Normal')
        
        # Statistics
        skewness = stats.skew(df[col])
        kurtosis = stats.kurtosis(df[col])
        axes[i].text(0.05, 0.95, f'Skew: {skewness:.2f}\nKurtosis: {kurtosis:.2f}', 
                    transform=axes[i].transAxes, fontsize=10, verticalalignment='top',
                    bbox=dict(boxstyle='round', facecolor='wheat', alpha=0.5))
        
        axes[i].set_title(f'{col.upper()} Distribution', fontsize=12, fontweight='bold')
        axes[i].set_xlabel(col)
        axes[i].set_ylabel('Density')
        axes[i].legend()
        axes[i].grid(True, alpha=0.3)
    
    plt.tight_layout()
    plt.savefig('../figures/distribution_analysis.png', dpi=300, bbox_inches='tight')
    plt.close()

def relationship_analysis(df):
    """Analyze relationships between variables"""
    print("=" * 80)
    print("RELATIONSHIP ANALYSIS")
    print("=" * 80)
    
    # Correlation matrix
    numerical_cols = ['age_years', 'height', 'weight', 'ap_hi', 'ap_lo', 'bmi', 'cardio']
    correlation_matrix = df[numerical_cols].corr()
    
    plt.figure(figsize=(12, 10))
    sns.heatmap(correlation_matrix, annot=True, cmap='coolwarm', center=0, 
                square=True, linewidths=1, cbar_kws={'shrink': 0.8}, fmt='.3f')
    plt.title('Correlation Matrix', fontsize=14, fontweight='bold')
    plt.tight_layout()
    plt.savefig('../figures/relationship_correlation.png', dpi=300, bbox_inches='tight')
    plt.close()
    
    # Relationship with target variable
    print("\nCorrelation with target variable (cardio):")
    cardio_corr = correlation_matrix['cardio'].sort_values(ascending=False)
    print(cardio_corr)

def target_variable_analysis(df):
    """Analyze target variable distribution and relationships"""
    print("=" * 80)
    print("TARGET VARIABLE ANALYSIS")
    print("=" * 80)
    
    # Distribution
    plt.figure(figsize=(10, 6))
    cardio_counts = df['cardio'].value_counts()
    plt.bar(cardio_counts.index.astype(str), cardio_counts.values, color=['skyblue', 'salmon'], edgecolor='black')
    plt.xlabel('Cardiovascular Disease', fontsize=12)
    plt.ylabel('Count', fontsize=12)
    plt.title('Distribution of Target Variable (Cardio)', fontsize=14, fontweight='bold')
    plt.xticks([0, 1], ['No (0)', 'Yes (1)'])
    for i, v in enumerate(cardio_counts.values):
        plt.text(i, v + 500, f'{v}\n({v/len(df)*100:.2f}%)', ha='center', va='bottom', fontweight='bold')
    plt.grid(True, alpha=0.3, axis='y')
    plt.tight_layout()
    plt.savefig('../figures/target_distribution.png', dpi=300, bbox_inches='tight')
    plt.close()
    
    # Comparison by target variable
    numerical_cols = ['age_years', 'height', 'weight', 'ap_hi', 'ap_lo', 'bmi']
    fig, axes = plt.subplots(2, 3, figsize=(18, 12))
    axes = axes.ravel()
    
    for i, col in enumerate(numerical_cols):
        df.boxplot(column=col, by='cardio', ax=axes[i])
        axes[i].set_title(f'{col.upper()} by Cardio Status', fontsize=12, fontweight='bold')
        axes[i].set_xlabel('Cardiovascular Disease')
        axes[i].set_ylabel(col)
        axes[i].grid(True, alpha=0.3)
    
    plt.suptitle('')
    plt.tight_layout()
    plt.savefig('../figures/target_comparison.png', dpi=300, bbox_inches='tight')
    plt.close()

def feature_engineering_insights(df):
    """Provide insights for feature engineering"""
    print("=" * 80)
    print("FEATURE ENGINEERING INSIGHTS")
    print("=" * 80)
    
    # Blood pressure categories
    df['bp_category'] = pd.cut(df['ap_hi'], bins=[0, 120, 140, 160, 300], 
                               labels=['Normal', 'Elevated', 'High Stage 1', 'High Stage 2'])
    print("\nBlood Pressure Categories:")
    print(df['bp_category'].value_counts())
    
    # BMI categories
    df['bmi_category'] = pd.cut(df['bmi'], bins=[0, 18.5, 25, 30, 100], 
                                labels=['Underweight', 'Normal', 'Overweight', 'Obese'])
    print("\nBMI Categories:")
    print(df['bmi_category'].value_counts())
    
    # Age groups
    df['age_group'] = pd.cut(df['age_years'], bins=[0, 30, 40, 50, 60, 100], 
                             labels=['<30', '30-40', '40-50', '50-60', '60+'])
    print("\nAge Groups:")
    print(df['age_group'].value_counts())

if __name__ == "__main__":
    df = load_and_clean_data()
    data_overview(df)
    outlier_analysis(df)
    distribution_analysis(df)
    relationship_analysis(df)
    target_variable_analysis(df)
    feature_engineering_insights(df)
    print("\nComprehensive EDA completed successfully!")

File: Comprehensive Health Data Analytics for Chronic Disease Prediction/r_scripts/comprehensive_eda.R
# Comprehensive Exploratory Data Analysis
# Deep dive into the health dataset

# Load libraries
library(ggplot2)
library(dplyr)
library(corrplot)
library(VIM)
library(psych)

# Install moments package if not available
if (!require(moments)) {
  install.packages("moments")
  library(moments)
}

# Load data
df <- read.csv("../data/health_data.csv", stringsAsFactors = FALSE)

# Data cleaning
if ("Unnamed..0" %in% colnames(df)) {
  df <- df[, !colnames(df) %in% "Unnamed..0"]
}

df$age_years <- df$age / 365.25
df$bmi <- df$weight / ((df$height / 100) ^ 2)
df <- df[df$ap_hi >= 80 & df$ap_hi <= 250, ]
df <- df[df$ap_lo >= 40 & df$ap_lo <= 150, ]
df <- df[df$ap_hi >= df$ap_lo, ]
df <- df[df$height >= 100 & df$height <= 220, ]
df <- df[df$weight >= 30 & df$weight <= 200, ]
df <- df[df$bmi >= 10 & df$bmi <= 60, ]

# ============================================================================
# DATA OVERVIEW
# ============================================================================

cat("=", rep("=", 78), "\n", sep="")
cat("DATA OVERVIEW\n")
cat("=", rep("=", 78), "\n", sep="")
cat("Dataset shape:", dim(df), "\n")
cat("\nColumn names:", colnames(df), "\n")
cat("\nData types:\n")
print(sapply(df, class))
cat("\nMissing values:\n")
print(colSums(is.na(df)))
cat("\nDuplicate rows:", sum(duplicated(df)), "\n")
cat("\nFirst few rows:\n")
print(head(df))
cat("\nSummary statistics:\n")
print(summary(df))

# Missing data visualization
if (sum(is.na(df)) > 0) {
  png("../figures/missing_data.png", width = 1200, height = 800, res = 300)
  aggr(df, col = c('steelblue', 'red'), numbers = TRUE, sortVars = TRUE,
       labels = names(df), cex.axis = 0.7, gap = 3,
       ylab = c("Missing data", "Pattern"))
  dev.off()
}

# ============================================================================
# OUTLIER ANALYSIS
# ============================================================================

cat("\n", "=", rep("=", 78), "\n", sep="")
cat("OUTLIER ANALYSIS\n")
cat("=", rep("=", 78), "\n", sep="")

numerical_cols <- c("age_years", "height", "weight", "ap_hi", "ap_lo", "bmi")

png("../figures/outlier_analysis.png", width = 1800, height = 1200, res = 300)
par(mfrow = c(2, 3))
for (col in numerical_cols) {
  Q1 <- quantile(df[[col]], 0.25, na.rm = TRUE)
  Q3 <- quantile(df[[col]], 0.75, na.rm = TRUE)
  IQR <- Q3 - Q1
  lower_bound <- Q1 - 1.5 * IQR
  upper_bound <- Q3 + 1.5 * IQR
  
  outliers <- sum(df[[col]] < lower_bound | df[[col]] > upper_bound, na.rm = TRUE)
  cat("\n", toupper(col), ":\n", sep="")
  cat("  Outliers:", outliers, "(", round(outliers/nrow(df)*100, 2), "%)\n")
  cat("  Lower bound:", lower_bound, "\n")
  cat("  Upper bound:", upper_bound, "\n")
  
  boxplot(df[[col]], main = paste(toupper(col), "- Box Plot"),
          ylab = col, col = "steelblue", border = "black")
  grid(nx = NA, ny = NULL)
}
dev.off()

# ============================================================================
# DISTRIBUTION ANALYSIS
# ============================================================================

cat("\n", "=", rep("=", 78), "\n", sep="")
cat("DISTRIBUTION ANALYSIS\n")
cat("=", rep("=", 78), "\n", sep="")

png("../figures/distribution_analysis.png", width = 1800, height = 1200, res = 300)
par(mfrow = c(2, 3))
for (col in numerical_cols) {
  hist(df[[col]], breaks = 50, probability = TRUE,
       main = paste(toupper(col), "Distribution"),
       xlab = col, col = "steelblue", border = "black")
  
  # Normal distribution overlay
  x <- seq(min(df[[col]], na.rm = TRUE), max(df[[col]], na.rm = TRUE), length = 100)
  lines(x, dnorm(x, mean = mean(df[[col]], na.rm = TRUE), 
                 sd = sd(df[[col]], na.rm = TRUE)), col = "red", lwd = 2)
  
  # Statistics
  if (require(moments, quietly = TRUE)) {
    skewness_val <- skewness(df[[col]], na.rm = TRUE)
    kurtosis_val <- kurtosis(df[[col]], na.rm = TRUE)
  } else {
    skewness_val <- NA
    kurtosis_val <- NA
  }
  legend("topright", 
         legend = c(paste("Skew:", round(skewness_val, 2)),
                   paste("Kurtosis:", round(kurtosis_val, 2))),
         bty = "n", bg = "wheat")
  grid()
}
dev.off()

# ============================================================================
# RELATIONSHIP ANALYSIS
# ============================================================================

cat("\n", "=", rep("=", 78), "\n", sep="")
cat("RELATIONSHIP ANALYSIS\n")
cat("=", rep("=", 78), "\n", sep="")

correlation_cols <- c("age_years", "height", "weight", "ap_hi", "ap_lo", "bmi", "cardio")
correlation_matrix <- cor(df[correlation_cols], use = "complete.obs")

png("../figures/relationship_correlation.png", width = 1200, height = 1000, res = 300)
corrplot(correlation_matrix, method = "color", type = "upper",
         order = "hclust", tl.cex = 0.8, tl.col = "black",
         addCoef.col = "black", number.cex = 0.7,
         title = "Correlation Matrix")
dev.off()

cat("\nCorrelation with target variable (cardio):\n")
cardio_corr <- sort(correlation_matrix["cardio", ], decreasing = TRUE)
print(cardio_corr)

# ============================================================================
# TARGET VARIABLE ANALYSIS
# ============================================================================

cat("\n", "=", rep("=", 78), "\n", sep="")
cat("TARGET VARIABLE ANALYSIS\n")
cat("=", rep("=", 78), "\n", sep="")

# Distribution
png("../figures/target_distribution.png", width = 1000, height = 600, res = 300)
cardio_counts <- table(df$cardio)
barplot(cardio_counts, names.arg = c("No (0)", "Yes (1)"),
        col = c("skyblue", "salmon"), border = "black",
        main = "Distribution of Target Variable (Cardio)",
        xlab = "Cardiovascular Disease", ylab = "Count")
text(barplot(cardio_counts, plot = FALSE), cardio_counts + 500,
     labels = paste0(cardio_counts, "\n(", round(cardio_counts/nrow(df)*100, 2), "%)"),
     pos = 3, font = 2)
grid(nx = NA, ny = NULL)
dev.off()

# Comparison by target variable
png("../figures/target_comparison.png", width = 1800, height = 1200, res = 300)
par(mfrow = c(2, 3))
for (col in numerical_cols) {
  boxplot(df[[col]] ~ df$cardio,
          main = paste(toupper(col), "by Cardio Status"),
          xlab = "Cardiovascular Disease", ylab = col,
          col = c("skyblue", "salmon"), border = "black")
  grid(nx = NA, ny = NULL)
}
dev.off()

# ============================================================================
# FEATURE ENGINEERING INSIGHTS
# ============================================================================

cat("\n", "=", rep("=", 78), "\n", sep="")
cat("FEATURE ENGINEERING INSIGHTS\n")
cat("=", rep("=", 78), "\n", sep="")

# Blood pressure categories
df$bp_category <- cut(df$ap_hi, breaks = c(0, 120, 140, 160, 300),
                      labels = c("Normal", "Elevated", "High Stage 1", "High Stage 2"))
cat("\nBlood Pressure Categories:\n")
print(table(df$bp_category))

# BMI categories
df$bmi_category <- cut(df$bmi, breaks = c(0, 18.5, 25, 30, 100),
                       labels = c("Underweight", "Normal", "Overweight", "Obese"))
cat("\nBMI Categories:\n")
print(table(df$bmi_category))

# Age groups
df$age_group <- cut(df$age_years, breaks = c(0, 30, 40, 50, 60, 100),
                    labels = c("<30", "30-40", "40-50", "50-60", "60+"))
cat("\nAge Groups:\n")
print(table(df$age_group))

cat("\nComprehensive EDA completed successfully!\n")

